skimage/util/_backends.py

```diff
@@ -0,0 +1,127 @@
+    then the scikit-image implementation is used.
+    """
+    func_name = func.__name__
+    func_module = func.__module__
+
+    # If no backends are installed at all or dispatching is disabled,
+    # return the original function. This way people who don't care about it
+    # don't see anything related to dispatching
+    if disable_dispatching() or not all_backends():
+        return func
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        # Backends are tried in alphabetical order, this makes things
+        # predictable and stable across runs. Might need a better solution
+        # when it becomes common that users have more than one backend
+        # that would accept a call.
+        for name in sorted(all_backends()):
+            backend = all_backends()[name]
+            # If we have a BackendInformation object we check if the
+            # function we are looking for is implemented in the backend
+            if "info" in backend:
+                if (
+                    f"{func_module}.{func_name}"
+                    not in backend["info"].supported_functions
+                ):
+                    continue
+
+            backend_impl = backend["implementation"].load()
+
+            # Allow the backend to accept/reject a call based on the function
+            # name and the values of the arguments
+            wants_it = backend_impl.can_has(
+                f"{func_module}.{func_name}", *args, **kwargs
+            )
+            if not wants_it:
+                continue
+
+            # can_has("foo", ...) might be True, but the function might not actually
+            # be implemented in the backend
+            func_impl = backend_impl.get_implementation(f"{func_module}.{func_name}")
+            if func_impl is not None:
+                warnings.warn(
+                    f"Call to '{func.__module__}.{func_name}' was dispatched to"
+                    f" the '{name}' backend. Set SKIMAGE_NO_DISPATCHING=1 to"
+                    " disable this.",
+                    DispatchNotification,
+                    # XXX from where should this warning originate?
+                    # XXX from where the function that was dispatched was called?
+                    # XXX or from where the user called a function that called
+                    # XXX a function that was dispatched?
+                    stacklevel=2,
+                )
+                return func_impl(*args, **kwargs)
+
+        else:
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+class BackendInformation:
+    """Information about a backend
+
+    A backend that wants to provide additional information about itself
+    should return an instance of this from its information entry-point.
+    """
+
+    def __init__(self, supported_functions):
+        self.supported_functions = supported_functions
+
+
+class DispatchNotification(RuntimeWarning):
```

Comment:
Why does this inherits from `RuntimeWarning`? This seems more closely related to logging messages than warning messages. But, it can be just my own personal bias I guess.

Reply:
That might be due to my suggestion. By default, I'd like users to be transparently informed that dispatching is enabled. I thought, Python's warning system would be a good fit for that, because it allows users to suppress this warning in multiple ways. 

But thinking about it more, this has some problems. This wouldn't play nicely with setups that treat warnings as errors. Logging is out too, because by default, users wouldn't see logged information.

So maybe we want to use a basic print statement that says something like "Hey we found valid backends and might use those by default with your current configuration". But how do we allow users to suppress this message if they don't want it? An environment variable?

Reply:
We did discuss `DispatchWarning` vs the slightly more unconventional `DispatchNotification` and the feeling was that if we called it `DispatchWarning` it would communicate to users that "something undesirable just happened". Which isn't the vibe we want. Hence using `DispatchNotification`.

Reply:
> I think the same warning from the same line of code won't be shown more than once?

Yes, so long you set the correct `stacklevel=` when giving the warning.  This is a pretty cool property indeed.

I don't necessary love warnings for their lack of thread-safety (unless that changed?) and the fact that I don't think they are particularly fast when hidden.

But, I suppose even if the future would default to not giving warnings (or hiding them by default), adding an option that disables the warnings (rather than relying on warning suppression) is not a big deal.

Reply:
Then let's try the warning route (for now) and collect experience with it.